File: programs/chronos/client/src/instruction/daemonInvoke.ts
import { Program } from "@project-serum/anchor";
import { PublicKey, TransactionInstruction } from "@solana/web3.js";
import { buildInstructionData, buildRemainingAccounts } from "./utils";
import { Chronos } from "../idl";
import { Account } from "../account";

export type DaemonInvokeArgs = {
  daemon: PublicKey;
  instruction: TransactionInstruction;
};

export class DaemonInvoke {
  private account: Account;
  private chronos: Program<Chronos>;

  constructor(account: Account, chronos: Program<Chronos>) {
    this.account = account;
    this.chronos = chronos;
  }

  public async daemonInvoke({
    daemon,
    instruction,
  }: DaemonInvokeArgs): Promise<TransactionInstruction> {
    const daemonData = await this.account.daemon.data(daemon);
    const instructionData = buildInstructionData(instruction);
    return this.chronos.instruction.daemonInvoke(instructionData, {
      accounts: {
        daemon: daemon,
        owner: daemonData.owner,
      },
      remainingAccounts: buildRemainingAccounts(instructionData, daemon),
    });
  }
}

File: src/instruction/daemon_invoke.rs
use anchor_client::anchor_lang::InstructionData;
use cronos_program::state::InstructionData as CronosInstructionData;
use solana_program::{
    instruction::{AccountMeta, Instruction},
    pubkey::Pubkey,
};

pub fn daemon_invoke(daemon: Pubkey, owner: Pubkey, instruction: Instruction) -> Instruction {
    Instruction {
        program_id: cronos_program::ID,
        accounts: vec![
            AccountMeta::new(daemon, false),
            AccountMeta::new(owner, true),
        ],
        data: cronos_program::instruction::DaemonInvoke {
            instruction_data: CronosInstructionData::from(instruction),
        }
        .data(),
    }
}
